File: src/python/finished/count_elements_with_maximum_frequency.py
# https://leetcode.com/problems/count-elements-with-maximum-frequency/

from collections import Counter
from typing import List


class Solution:
    def maxFrequencyElements(self, nums: List[int]) -> int:
        cnt = Counter(nums)

        cur_freq = 0
        cur_sum = 0
        for freq in cnt.values():
            if freq < cur_freq:
                continue
            if freq > cur_freq:
                cur_freq = freq
                cur_sum = freq
                continue
            cur_sum += freq

        return cur_sum

File: src/rust/finished/count_elements_with_maximum_frequency.rs
impl Solution {
    pub fn max_frequency_elements(nums: Vec<i32>) -> i32 {
        let freq = nums
            .iter()
            .fold(std::collections::HashMap::new(), |mut prev, n| {
                match prev.get_mut(n) {
                    Some(v) => *v += 1,
                    None => {
                        prev.insert(n, 1);
                    }
                };
                prev
            });

        freq.values()
            .fold((0, 0), |(cur_freq, cur_sum), &freq| match cur_freq - freq {
                ..0 => (freq, freq),
                0 => (cur_freq, cur_sum + freq),
                1.. => (cur_freq, cur_sum),
            })
            .1
    }
}
